tpm/ak.go

```diff
@@ -0,0 +1,507 @@
+// Names. It returns `ErrNotFound` if it doesn't exist.
+func (t *TPM) GetAKByPermanentIdentifier(ctx context.Context, permanentIdentifier string) (*AK, error) {
+	var err error
+	if err = t.open(ctx); err != nil {
+		return nil, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer func() {
+		if tempErr := t.close(ctx); tempErr != nil && err != nil {
+			err = tempErr
+		}
+	}()
+
+	aks, err := t.ListAKs(internalCall(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	// loop through all available AKs and check if one exist that
+	// contains a Subject Alternative Name extension containing the
+	// requested PermanentIdentifier.
+	for _, ak := range aks {
+		if ak.HasValidPermanentIdentifier(permanentIdentifier) {
+			return ak, nil
+		}
+	}
+
+	return nil, ErrNotFound
+}
+
+// ListAKs returns a slice of AKs. The result is (currently)
+// not ordered.
+func (t *TPM) ListAKs(ctx context.Context) ([]*AK, error) {
+	var err error
+	if err := t.open(ctx); err != nil {
+		return nil, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer func() {
+		if tempErr := t.close(ctx); tempErr != nil && err != nil {
+			err = tempErr
+		}
+	}()
+
+	aks, err := t.store.ListAKs()
+	if err != nil {
+		return nil, fmt.Errorf("failed listing AKs: %w", err)
+	}
+
+	result := make([]*AK, 0, len(aks))
+	for _, ak := range aks {
+		result = append(result, akFromStorage(ak, t))
+	}
+
+	// TODO: include ordering by name or createdAt?
+
+	return result, nil
+}
+
+// DeleteAK removes the AK identified by `name`. It returns `ErrNotfound`
+// if it doesn't exist. Keys that were attested by the AK have to be removed
+// before removing the AK, otherwise an error will be returned.
+func (t *TPM) DeleteAK(ctx context.Context, name string) error {
+	var err error
+	if err := t.open(ctx); err != nil {
+		return fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer func() {
+		if tempErr := t.close(ctx); tempErr != nil && err != nil {
+			err = tempErr
+		}
+	}()
+
+	ak, err := t.store.GetAK(name)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return fmt.Errorf("failed getting AK %q: %w", name, ErrNotFound)
+		}
+		return fmt.Errorf("failed getting AK %q: %w", name, err)
+	}
+
+	// prevent deleting the AK if the TPM (storage) contains keys that
+	// were attested by it. While keys would still work if the AK were
+	// deleted, some functionalities would no longer work. The AK can
+	// only be deleted if all keys attested by it are deleted first.
+	keys, err := t.GetKeysAttestedBy(internalCall(ctx), name)
+	if err != nil {
+		return fmt.Errorf("failed getting keys attested by AK %q: %w", name, err)
+	}
+
+	if len(keys) > 0 {
+		return fmt.Errorf("cannot delete AK %q before deleting keys that were attested by it", name)
+	}
+
+	if err := t.attestTPM.DeleteKey(ak.Data); err != nil { // TODO: we could add a DeleteAK to go-attestation; under the hood it's loaded the same as a key though.
+		return fmt.Errorf("failed deleting AK %q: %w", name, err)
+	}
```

Comment:
On Linux the call doesn't do anything currently, because the TPM doesn't store the key. On Windows we might get an error like `0x8028000D: "TPM_E_KEYNOTFOUND"`, when doing `nCryptDeleteKey`. The full error would look something like this: `nCryptDeleteKey returned 8028000D: tpm or subsystem failure: TPM_E_KEYNOTFOUND`. We can only rely on checking the string, currently. 

Are you OK with a follow-up PR for this? I've made a note of this.

Reply:
👍